kvserver/concurrency: TestConcurrencyManagerBasic/discovered_lock flaked

https://teamcity.cockroachdb.com/viewLog.html?buildId=1789025&tab=buildResultsDiv&buildTypeId=Cockroach_UnitTests_Test

```
------- Stdout: -------
=== RUN   TestConcurrencyManagerBasic/discovered_lock

testdata/concurrency_manager/discovered_lock:6:
new-txn [3 args]
<no input to command>
----

testdata/concurrency_manager/discovered_lock:9:
new-txn [3 args]
<no input to command>
----

testdata/concurrency_manager/discovered_lock:12:
new-request [3 args]
get key=k
----

testdata/concurrency_manager/discovered_lock:16:
sequence [1 args]
<no input to command>
----
[1] sequence req1: sequencing request
[1] sequence req1: acquiring latches
[1] sequence req1: scanning lock table for conflicting locks
[1] sequence req1: sequencing complete, returned guard

testdata/concurrency_manager/discovered_lock:23:
handle-write-intent-error [3 args]
<no input to command>
----
[-] handle write intent error req1: handling conflicting intents on "k"

testdata/concurrency_manager/discovered_lock:27:
debug-lock-table [0 args]
<no input to command>
----
global: num=1
 lock: "k"
  holder: txn: 00000001-0000-0000-0000-000000000000, ts: 0.000000010,1
local: num=0
    --- FAIL: TestConcurrencyManagerBasic/discovered_lock (0.03s)
        datadriven.go:165: 
            testdata/concurrency_manager/discovered_lock:34: 
            expected:
            [2] sequence req1: re-sequencing request
            [2] sequence req1: acquiring latches
            [2] sequence req1: scanning lock table for conflicting locks
            [2] sequence req1: waiting in lock wait-queues
            [2] sequence req1: pushing txn 00000001
            [2] sequence req1: blocked on select in concurrency_test.(*cluster).PushTransaction
            
            found:
            [2] sequence req1: re-sequencing request
            [2] sequence req1: acquiring latches
            [2] sequence req1: scanning lock table for conflicting locks
            [2] sequence req1: waiting in lock wait-queues
            [2] sequence req1: pushing txn 00000001
            [2] sequence req1: blocked on select in concurrency.(*lockTableWaiterImpl).WaitOn
```

--- a/pkg/kv/kvserver/concurrency/lock_table_waiter.go
+++ b/pkg/kv/kvserver/concurrency/lock_table_waiter.go
@@ -187,12 +187,21 @@ func (w *lockTableWaiterImpl) WaitOn(
 					delay = 0
 				}
 
-				if timer == nil {
-					timer = timeutil.NewTimer()
-					defer timer.Stop()
+				if delay > 0 {
+					if timer == nil {
+						timer = timeutil.NewTimer()
+						defer timer.Stop()
+					}
+					timer.Reset(delay)
+					timerC = timer.C
+				} else {
+					// If we don't want to delay the push, don't use a real timer.
+					// Doing so is both a waste of resources and, more importantly,
+					// makes TestConcurrencyManagerBasic flaky because there's no
+					// guarantee that the timer will fire before the goroutine enters
+					// a "select" waiting state on the next iteration of this loop.
+					timerC = closedTimerC
 				}
-				timer.Reset(delay)
-				timerC = timer.C
 				timerWaitingState = state
 
 			case waitElsewhere:
@@ -237,7 +246,9 @@ func (w *lockTableWaiterImpl) WaitOn(
 			// of a dependency cycle or that the lock holder's coordinator node
 			// has crashed.
 			timerC = nil
-			timer.Read = true
+			if timer != nil {
+				timer.Read = true
+			}
 
 			// If the request is conflicting with a held lock then it pushes its
 			// holder synchronously - there is no way it will be able to proceed
@@ -450,4 +461,11 @@ func minDuration(a, b time.Duration) time.Duration {
 		return a
 	}
 	return b
+}
+
+var closedTimerC chan time.Time
+
+func init() {
+	closedTimerC = make(chan time.Time)
+	close(closedTimerC)
 }

